src/libmongoc/tests/unified/entity-map.c

```diff
@@ -0,0 +1,603 @@
+      BSON_FOREACH (observe_events, iter)
+      {
+         const char *event_type = bson_iter_utf8 (&iter, NULL);
+
+         if (0 == strcmp (event_type, "commandStartedEvent")) {
+            mongoc_apm_set_command_started_cb (callbacks, command_started);
+         } else if (0 == strcmp (event_type, "commandFailedEvent")) {
+            mongoc_apm_set_command_failed_cb (callbacks, command_failed);
+         } else if (0 == strcmp (event_type, "commandSucceededEvent")) {
+            mongoc_apm_set_command_succeeded_cb (callbacks, command_succeeded);
+         } else {
+            test_set_error (error, "Unexpected event type: %s", event_type);
+            goto done;
+         }
+      }
+   }
+   mongoc_client_set_apm_callbacks (client, callbacks, entity);
+
+   ret = true;
+done:
+   mongoc_uri_destroy (uri);
+   bson_parser_destroy (parser);
+   mongoc_apm_callbacks_destroy (callbacks);
+   bson_destroy (uri_options);
+   bson_free (use_multiple_mongoses);
+   bson_destroy (observe_events);
+   if (!ret) {
+      entity_destroy (entity);
+      return NULL;
+   }
+   return entity;
+}
+
+entity_t *
+entity_database_new (entity_map_t *entity_map,
+                     bson_t *bson,
+                     bson_error_t *error)
+{
+   bson_parser_t *parser;
+   entity_t *entity;
+   const entity_t *client_entity;
+   char *client_id = NULL;
+   mongoc_client_t *client;
+   char *database_name = NULL;
+   bool ret = false;
+
+   entity = entity_new ("database");
+   parser = bson_parser_new ();
+   bson_parser_utf8 (parser, "id", &entity->id);
+   bson_parser_utf8 (parser, "client", &client_id);
+   bson_parser_utf8 (parser, "databaseName", &database_name);
+
+   if (!bson_parser_parse (parser, bson, error)) {
+      goto done;
+   }
+
+   client_entity = entity_map_get (entity_map, client_id, error);
+   if (!client_entity) {
+      goto done;
+   }
+
+   client = (mongoc_client_t *) client_entity->value;
+   entity->value = (void *) mongoc_client_get_database (client, database_name);
+
+   ret = true;
+done:
+   bson_free (client_id);
+   bson_free (database_name);
+   bson_parser_destroy (parser);
+   if (!ret) {
+      entity_destroy (entity);
+      return NULL;
+   }
+   return entity;
+}
+
+entity_t *
+entity_collection_new (entity_map_t *entity_map,
+                       bson_t *bson,
+                       bson_error_t *error)
+{
+   bson_parser_t *parser;
+   entity_t *entity;
+   entity_t *database_entity;
+   mongoc_database_t *database;
+   bool ret = false;
+   char *database_id = NULL;
+   char *collection_name = NULL;
+
+   entity = entity_new ("collection");
+   parser = bson_parser_new ();
+   bson_parser_utf8 (parser, "id", &entity->id);
+   bson_parser_utf8 (parser, "database", &database_id);
+   bson_parser_utf8 (parser, "collectionName", &collection_name);
+   if (!bson_parser_parse (parser, bson, error)) {
+      goto done;
+   }
+
+   database_entity = entity_map_get (entity_map, database_id, error);
```

Comment:
In theory we could fail this lookup if a test makes a mistake with its entity list or defines entities out-of-order, right?  Let's just add a check here that would help us debug.

Reply:
Oh yes, done. `error` will be set if `database_id` is not found.